src/libinjector/linux/methods/linux_read_file.c

```diff
@@ -0,0 +1,371 @@
+            break;
+        }
+        case STEP3: // verify fd and setups first read syscall
+        {
+            if ( is_syscall_error(info->regs->rax) )
+            {
+                fprintf(stderr, "Could not open file in guest\n");
+                return cleanup(drakvuf, info, true);
+            }
+
+            injector->fd = info->regs->rax;
+            PRINT_DEBUG("File descriptor: %ld\n", injector->fd);
+
+            if (!setup_read_syscall(injector, info->regs, FILE_BUF_SIZE))
+                PRINT_DEBUG("Failed to setup read syscall\n");
+
+            info->regs->rip = injector->syscall_addr;
+
+            event = VMI_EVENT_RESPONSE_SET_REGISTERS;
+            break;
+        }
+        case STEP4: // loop till all chunks are written and then close the fd
+        {
+            if ( is_syscall_error(info->regs->rax) )
+            {
+                fprintf(stderr, "Could not read chunk from guest\n");
+                return cleanup(drakvuf, info, true);
+            }
+            injector->buffer.len = info->regs->rax;
+
+            if (!injector->buffer.len)
+            {
+                PRINT_DEBUG("Read file successful\n");
+                PRINT_DEBUG("File size: (%ld)\n", injector->buffer.total_processed);
+
+                if (!setup_close_syscall(injector, info->regs))
+                {
+                    PRINT_DEBUG("Failed to setup close syscall\n");
+                    return cleanup(drakvuf, info, true);
+                }
+            }
+            else
+            {
+                PRINT_DEBUG("Chunk read successful (%ld)\n", injector->buffer.len);
+                injector->buffer.total_processed += injector->buffer.len;
+
+                if (!write_buffer_to_file(drakvuf, info, injector->buffer.len))
+                    return cleanup(drakvuf, info, true);
+
+                if (!setup_read_syscall(injector, info->regs, FILE_BUF_SIZE))
+                {
+                    PRINT_DEBUG("Failed to setup read syscall\n");
+                    return cleanup(drakvuf, info, true);
+                }
+
+                injector->step_override = true;
+                injector->step = STEP4;
+            }
+
+            info->regs->rip = injector->syscall_addr;
+
+            event = VMI_EVENT_RESPONSE_SET_REGISTERS;
+            break;
+        }
+        case STEP5: // restore the registers
+        {
+            PRINT_DEBUG("Closed File descriptor\n");
+            PRINT_DEBUG("Restoring state\n");
+            free_bp_trap(drakvuf, injector, info->trap);
+
+            // restore regs
+            memcpy(info->regs, &injector->saved_regs, sizeof(x86_registers_t));
+
+            event = VMI_EVENT_RESPONSE_SET_REGISTERS;
+            break;
+        }
+        case STEP6: // cleanup
+        {
+            PRINT_DEBUG("Removing traps and exiting\n");
+
+            // remove the initial trap here
+            free_bp_trap(drakvuf, injector, info->trap);
```

Comment:
I'm worried about these 2 `free`s, as they implicitly cast `info->trap` to bool.

Reply:
> as they implicitly cast info->trap to bool.

Umm, I am not sure I understand this. Could you explain it more with this on where it might go wrong?
https://github.com/tklengyel/drakvuf/blob/58529dc52c1ab4f0ab08ecd4b2061188fa12098a/src/libinjector/linux/linux_utils.c#L274-L278

Reply:
I don't see anything wrong here either